consumer/consumererror/error.go

```diff
@@ -0,0 +1,179 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package consumererror // import "go.opentelemetry.io/collector/consumer/consumererror"
+
+import (
+	"net/http"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"go.opentelemetry.io/collector/consumer/consumererror/internal/statusconversion"
+)
+
+// Error is intended to be used to encapsulate various information that can add
+// context to an error that occurred within a pipeline component. Error objects
+// are constructed through calling `New` with the relevant options to capture
+// data around the error that occurred.
+//
+// It may hold multiple errors from downstream components, and can be merged
+// with other errors as it travels upstream using `Combine`. The `Error` should
+// be obtained from a given `error` object using `errors.As`.
+//
+// Experimental: This API is at the early stage of development and may change
+// without backward compatibility
+type Error struct {
+	error
+	httpStatus int
+	grpcStatus *status.Status
+	retryable  bool
+}
+
+var _ error = (*Error)(nil)
+
+// ErrorOption allows annotating an Error with metadata.
+type ErrorOption interface {
+	applyOption(*Error)
+}
+
+type errorOptionFunc func(*Error)
+
+func (f errorOptionFunc) applyOption(e *Error) {
+	f(e)
+}
+
+// New wraps an error that happened while consuming telemetry and adds metadata
+// onto it to be passed back up the pipeline.
+// At least one option should be provided.
+//
+// Experimental: This API is at the early stage of development and may change
+// without backward compatibility
+func New(origErr error, options ...ErrorOption) error {
+	err := &Error{error: origErr}
+
+	for _, option := range options {
+		option.applyOption(err)
+	}
+
+	return err
+}
+
+// WithOTLPHTTPStatus records an HTTP status code that was received from a server
+// during data submission.
+// It is not necessary to use WithRetryable with creating an error with WithOTLPHTTPStatus
+// as the retryable property can be inferred from the HTTP status code using OTLP specification.
+//
+// Experimental: This API is at the early stage of development and may change
+// without backward compatibility
+func WithOTLPHTTPStatus(status int) ErrorOption {
+	return errorOptionFunc(func(err *Error) {
+		err.httpStatus = status
+	})
+}
+
+// WithOTLPGRPCStatus records a gRPC status code that was received from a server
+// during data submission.
+// It is not necessary to use WithRetryable with creating an error with WithOTLPGRPCStatus
+// as the retryable property can be inferred from the grpc status using OTLP specification.
+//
+// Experimental: This API is at the early stage of development and may change
+// without backward compatibility
+func WithOTLPGRPCStatus(status *status.Status) ErrorOption {
```

Comment:
I still don't see a reason to have options.  I don't think there is a need to have an `Error` that can be initiated as a combination of gRPC, HTTP and internal collector error. Any error can be originated only from one of those three categories. I believe we need constructors for each of them individually, not options. AFAIR, we agreed on that during the call. cc @bogdandrutu

Reply:
Options removed.